### PR TITLE
Chore/rewrite events to promises

### PR DESCRIPTION
**Pull request by MarkNerdi996**
## Summary

> Please summarize your changes, describing **what** they are and **why** they were made.

...

## Changelog

```
- Please write a list of changes
```

### Platforms

> Please select any platforms where your changes have been tested.

-   **Desktop**
    -   [ ] MacOS
    -   [ ] Linux
    -   [ ] Windows
-   **Mobile**
    -   [ ] iOS
    -   [ ] Android

### Instructions

> Please describe the specific instructions, configurations, and/or test cases necessary to **test and verify** that your changes work as intended.

...

## Checklist

> Please tick the following boxes that are relevant to your changes.

-   [ ] I have followed the contribution guidelines for this project
-   [ ] I have performed a self-review of my own code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have added or modified tests that prove my changes work as intended
-   [ ] I have verified that new and existing unit tests pass locally with my changes
-   [ ] I have verified that my latest changes pass CI workflows for testing and linting
-   [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: packages/shared/lib/core/layer-2/utils/signIscpTransferTransactionData.ts [hunk 1]
```diff
@@ -11,44 +10,44 @@ import {
...
-            Ledger.signEvmTransaction(transaction, bip32)
-        } else {
-            throw new Error('Unable to find web3 provider.')
-        }
-    } catch (err) {
-        console.error(err)
+): Promise<string | undefined> {
+    const chain = getNetwork()?.getChain(asset.chainId)
+    if (!chain) {
+        return
```
Comment (Tuditi) [inline packages/shared/lib/core/layer-2/utils/signIscpTransferTransactionData.ts:23]: Throw an error here as well?

File: packages/desktop/electron/lib/ledger.js [hunk 2]
```diff
@@ -80,20 +80,23 @@ async function signTransactionData(data, bip32Path) {
...
+            TX_OPTIONS
+        )
+
+        const serializedSignedTransaction = Buffer.from(RLP.encode(bufArrToArr(signedTransactionObject.raw())))
+        const serializedSignedTransactionString = '0x' + serializedSignedTransaction.toString('hex')
+
+        return { signedTransaction: serializedSignedTransactionString }
+    } catch (error) {
+        return { signedTransaction: undefined }
+    }
```
Comment (Tuditi) [inline packages/desktop/electron/lib/ledger.js:101]: Try/catching here defeats the purpose of the try/catch statement on l.28 above.

File: packages/shared/lib/core/layer-2/utils/signTransactionWithLedger.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        if (!isSigning) {
+            if (signedTransaction) {
+                return Promise.resolve(signedTransaction)
+            } else {
+                return Promise.reject('Rejected')
+            }
+        }
+        await sleep(pollingInterval)
+    }
+    return Promise.reject('Timeout')
```
Comment (Tuditi) [inline packages/shared/lib/core/layer-2/utils/signTransactionWithLedger.ts:35]: This works very nicely. I think we should replace the error messages with locale's. The error on l.30 can say something like: `You must open the Ethereum app on your Ledger device`?

The error on l.35 could use the `error.ledger.temout` locale.

**Review by Tuditi — CHANGES_REQUESTED** [review]

File: packages/desktop/electron/lib/ledger.js (hunk 2 above)
Comment (MarkNerdi996) [reply packages/desktop/electron/lib/ledger.js:101]: the use for try catch is that if you `Reject` on the ledger device, we want to return undefined to the renderer process, such that we can handle rejecting there

File: packages/shared/lib/core/layer-2/utils/signTransactionWithLedger.ts (hunk 3 above)
Comment (MarkNerdi996) [reply packages/shared/lib/core/layer-2/utils/signTransactionWithLedger.ts:35]: the error on line 30 will be thrown if the user `Rejects` the signing. So it should rather be `Signing was rejected by the Ledger device`.

Comment (Tuditi) [reply packages/shared/lib/core/layer-2/utils/signTransactionWithLedger.ts]: It's also thrown when the wrong app is open, but I'll add fixing this as a polish task. Ideally we should use Matt's function to check which app is open and show a pop-up if it's the wrong one.

File: packages/desktop/electron/lib/ledger.js (hunk 2 above)
Comment (Tuditi) [reply packages/desktop/electron/lib/ledger.js:101]: Aaah, okay, I see now, where you use this. Maybe we can improve this by sending back the error, so we can make a distinction between the user rejecting or the wrong app being opened on the device. In any case, it's outside the scope of this PR.

**Review by Tuditi — APPROVED** [review]
LGTM! :fire:

**Pull request merged**